apps/aeoracle/src/aeo_query_tx.erl

```diff
@@ -0,0 +1,202 @@
+-module(aeo_query_tx).
+
+-include("oracle_txs.hrl").
+-include_lib("apps/aecore/include/trees.hrl").
+
+-behavior(aetx).
+
+%% Behavior API
+-export([new/1,
+         fee/1,
+         nonce/1,
+         origin/1,
+         check/3,
+         process/3,
+         signers/1,
+         serialize/1,
+         deserialize/1,
+         type/0
+        ]).
+
+%% Additional getters
+-export([oracle/1,
+         query/1,
+         query_fee/1,
+         query_ttl/1,
+         response_ttl/1,
+         sender/1]).
+
+
+-define(ORACLE_QUERY_TX_TYPE, <<"oracle_query">>).
+-define(ORACLE_QUERY_TX_VSN, 1).
+-define(ORACLE_QUERY_TX_FEE, 2).
+
+-opaque query_tx() :: #oracle_query_tx{}.
+
+-export_type([query_tx/0]).
+
+-spec sender(query_tx()) -> pubkey().
+sender(#oracle_query_tx{sender = SenderPubKey}) ->
+    SenderPubKey.
+
+-spec oracle(query_tx()) -> pubkey().
+oracle(#oracle_query_tx{oracle = OraclePubKey}) ->
+    OraclePubKey.
+
+-spec query(query_tx()) -> string().
+query(#oracle_query_tx{query = Query}) ->
+    Query.
+
+-spec query_fee(query_tx()) -> integer().
+query_fee(#oracle_query_tx{query_fee = QueryFee}) ->
+    QueryFee.
+
+-spec query_ttl(query_tx()) -> aeo_oracles:ttl().
+query_ttl(#oracle_query_tx{query_ttl = QueryTTL}) ->
+    QueryTTL.
+
+-spec response_ttl(query_tx()) -> aeo_oracles:relative_ttl().
+response_ttl(#oracle_query_tx{response_ttl = ResponseTTL}) ->
+    ResponseTTL.
+
+-spec new(map()) -> {ok, query_tx()}.
+new(#{sender        := SenderPubKey,
+      nonce         := Nonce,
+      oracle        := Oracle,
+      query         := Query,
+      query_fee     := QueryFee,
+      query_ttl     := QueryTTL,
+      response_ttl  := ResponseTTL,
+      fee           := Fee}) ->
+    {ok, #oracle_query_tx{sender        = SenderPubKey,
+                          nonce         = Nonce,
+                          oracle        = Oracle,
+                          query         = Query,
+                          query_fee     = QueryFee,
+                          query_ttl     = QueryTTL,
+                          response_ttl  = ResponseTTL,
+                          fee           = Fee}}.
+
+-spec fee(query_tx()) -> integer().
+fee(#oracle_query_tx{fee = F}) ->
+    F.
+
+-spec nonce(query_tx()) -> non_neg_integer().
+nonce(#oracle_query_tx{nonce = Nonce}) ->
+    Nonce.
+
+-spec origin(query_tx()) -> pubkey().
+origin(#oracle_query_tx{sender = SenderPubKey}) ->
+    SenderPubKey.
+
+%% SenderAccount should exist, and have enough funds for the fee + the query_fee.
+%% Oracle should exist, and query_fee should be enough
+%% Fee should cover TTL
+%% TODO: Check the query? Check TTL relative to Oracle TTL?
+-spec check(query_tx(), trees(), height()) -> {ok, trees()} | {error, term()}.
+check(#oracle_query_tx{sender = SenderPubKey, nonce = Nonce,
+                       oracle = OraclePubKey, query_fee = QFee,
+                       query_ttl = TTL, fee = Fee}, Trees, Height) ->
+    Checks =
```

Comment:
In processing I see insertion of new interaction `aeo_state_tree:enter_interaction` though I do not see check that interaction is not there. Am I misunderstanding model, or maybe validation shall be here?

Reply:
Maybe explicit validation would be adding clarity... Since the sender nonce is part of the Id of the interaction check_account would stop an attempt to use the same nonce twice...

Reply:
This should indeed be validated.

Reply:
Added validation.